File: chemistry/chemistry_mpi.f90
! -*- Mode: Fortran90; -*-
!-----------------------------------------------------------------
! Daniel R. Reynolds
! SMU, Mathematics
! Math 6370
! 9 January 2009
!=================================================================


program ChemicalEquilibrium_MPI
  !-----------------------------------------------------------------
  ! Description: 
  !    Computes the equilibrium chemical densities at a number of 
  !    spatial locations, given a (random) background temperature 
  !    field.  The chemical rate equations and solution strategy
  !    are in the subroutine chem_solver, which is called at every
  !    spatial location.
  !-----------------------------------------------------------------
  !======= Inclusions ===========
  use mpi
!  include "mpif.h"  ! include this if mpi module unavailable
  implicit none

  !======= Interfaces ===========
  interface
     subroutine chem_solver(T,u,v,w,lam,eps,maxit,its,res)
       integer, intent(in) :: maxit
       integer, intent(out) :: its
       double precision, intent(in) :: T, lam, eps
       double precision, intent(out) :: res
       double precision, intent(inout) :: u, v, w
     end subroutine chem_solver
  end interface

  !======= Declarations =========
  integer :: n, i, is, ie, js, je, maxit, its, numprocs, myid, ierr
  integer, allocatable :: counts(:), displs(:)
  double precision, allocatable :: T(:), u(:), v(:), w(:)
  double precision :: lam, eps, res, stime, ftime

  !======= Internals ============
  
  ! set solver input parameters
  maxit = 1000000
  lam = 1.d-2
  eps = 1.d-10
  
  ! intialize MPI
  call mpi_init(ierr)
  if (ierr /= 0) then
     write(*,*) ' error in mpi_init =',ierr
     stop
  endif
  call mpi_comm_size(MPI_COMM_WORLD, numprocs, ierr)
  if (ierr /= 0) then
     write(*,*) ' error in mpi_comm_size =',ierr
     call mpi_abort(MPI_COMM_WORLD, -1, ierr)
     stop
  endif
  call mpi_comm_rank(MPI_COMM_WORLD, myid, ierr)
  if (ierr /= 0) then
     write(*,*) ' error in mpi_comm_rank =',ierr
     call mpi_abort(MPI_COMM_WORLD, -1, ierr)
     stop
  endif

  ! input the total number of intervals
  if (myid == 0) then
     write(*,*) 'Enter the total number of intervals (0 quits):'
     read(*,*) n
  endif

  ! root node outputs parallelism information to screen
  if (myid == 0) then
     write(*,'(A,i5,A)') ' starting MPI with ', numprocs,' processes'
  endif

  ! root node sends n out to other processors
  call mpi_bcast(n, 1, MPI_INTEGER, 0, MPI_COMM_WORLD, ierr)
  if (ierr /= 0) then
     write(*,*) ' error in mpi_bcast =',ierr
     call mpi_abort(MPI_COMM_WORLD, -1, ierr)
     stop
  endif
  
  ! stop for illegal n
  if (n < 1) then
     call mpi_finalize(ierr)
     stop
  endif

  ! determine this processor's interval
  is = floor(1.d0*n/numprocs)*myid + 1
  ie = floor(1.d0*n/numprocs)*(myid+1)
  if (myid == numprocs-1) then
     ie = n
  endif

  ! root node allocates temperature field and solution arrays
  if (myid == 0) then
     allocate(T(n),u(n),v(n),w(n))

     ! set random temperature field, initial guesses at chemical densities
     call random_number(T)
     u = 0.35d0
     v = 0.1d0
     w = 0.5d0
  else  ! other nodes allocate local temperature field and solution arrays
     allocate(T(is:ie),u(is:ie),v(is:ie),w(is:ie))

     ! set initial guesses at chemical densities
     u = 0.35d0
     v = 0.1d0
     w = 0.5d0
  endif

  ! allocate gatherv/scatterv temporary arrays
  allocate(counts(numprocs),displs(numprocs))

  ! start timer
  stime = MPI_Wtime()

  ! root node sends out portions of temperature field to all procs
  do i=0,numprocs-1
     js = floor(1.d0*n/numprocs)*i + 1
     je = floor(1.d0*n/numprocs)*(i+1)
     if (i == numprocs-1) then
        je = n
     endif
     counts(i+1) = je-js+1
     displs(i+1) = js-1
  enddo
  call mpi_scatterv(T, counts, displs, MPI_DOUBLE_PRECISION, T, ie-is+1, &
                   MPI_DOUBLE_PRECISION, 0, MPI_COMM_WORLD, ierr)

  ! call solver over local intervals
  do i=is,ie,1
     call chem_solver(T(i),u(i),v(i),w(i),lam,eps,maxit,its,res)
     if (res < eps) then
!!$        write(*,'(2(A,i8))') '    i =',i,',  its =',its
     else
        write(*,'(2(A,i8),5(A,es9.2))') '    error: i =',i,',  its =',its, &
             ',  res =',res,',  T =',T(i),',  u =',u(i),',  v =',v(i),',  w =',w(i)
        call mpi_abort(MPI_COMM_WORLD, -1, ierr)
     endif
  enddo

  ! root node collects results
  call mpi_gatherv(u, ie-is+1, MPI_DOUBLE_PRECISION, u, counts, displs, &
                   MPI_DOUBLE_PRECISION, 0, MPI_COMM_WORLD, ierr)
  if (ierr /= 0) then
     write(*,*) ' error in mpi_gatherv =',ierr
     call mpi_abort(MPI_COMM_WORLD, -1, ierr)
     stop
  endif
  call mpi_gatherv(v, ie-is+1, MPI_DOUBLE_PRECISION, v, counts, displs, &
                   MPI_DOUBLE_PRECISION, 0, MPI_COMM_WORLD, ierr)
  if (ierr /= 0) then
     write(*,*) ' error in mpi_gatherv =',ierr
     call mpi_abort(MPI_COMM_WORLD, -1, ierr)
     stop
  endif
  call mpi_gatherv(w, ie-is+1, MPI_DOUBLE_PRECISION, w, counts, displs, &
                   MPI_DOUBLE_PRECISION, 0, MPI_COMM_WORLD, ierr)
  if (ierr /= 0) then
     write(*,*) ' error in mpi_gatherv =',ierr
     call mpi_abort(MPI_COMM_WORLD, -1, ierr)
     stop
  endif

  ! stop timer
  ftime = MPI_Wtime()

  ! output solution time
  write(*,'(A,i3,A,i5,A)') ' proc ',myid,' computed ',ie-is+1,' iterations'
  if (myid == 0) then
     write(*,*) '     runtime =',ftime-stime
  endif


  ! free temperature/solution arrays and mpi temporary arrays
  deallocate(T,u,v,w,counts,displs)

  ! finalize MPI
  call mpi_finalize(ierr)  

end program ChemicalEquilibrium_MPI
!=================================================================

File: chemistry/chemistry_mpi.c
/* Daniel R. Reynolds
   SMU Mathematics
   Math 6370
   20 January 2011 */


/* Inclusions */
#include <stdlib.h>
#include <stdio.h>
#include <math.h>
#include <stdint.h>
#include "mpi.h"


/* Prototypes */
void chem_solver(double, double*, double*, double*, 
		 double, double, int, int*, double*);


/* Example routine to compute the equilibrium chemical densities at 
   a number of spatial locations, given a (random) background temperature
   field.  The chemical rate equations and solution strategy are in the 
   subroutine chem_solver, which is called at every spatial location. */
int main(int argc, char* argv[]) {

  /* intialize MPI */
  int i, ierr, numprocs, myid;
  ierr = MPI_Init(&argc, &argv);
  if (ierr != MPI_SUCCESS) {
    printf("Error in MPI_Init = %i\n",ierr);
    return -1;
  }
  ierr = MPI_Comm_size(MPI_COMM_WORLD, &numprocs);
  if (ierr != MPI_SUCCESS) {
    printf("Error in MPI_Comm_size = %i\n",ierr);
    return -1;
  }
  ierr = MPI_Comm_rank(MPI_COMM_WORLD, &myid);
  if (ierr != MPI_SUCCESS) {
    printf("Error in MPI_Comm_rank = %i\n",ierr);
    return -1;
  }

  /* set solver input parameters */
  int maxit = 1000000;
  double lam = 1.e-2;
  double eps = 1.e-10;

  /* input the number of intervals */
  int n;
  if (myid == 0) {
    printf("Enter the total number of intervals (0 quits):\n");
    i = scanf("%i", &n);
    if (i < 1) {
      ierr = MPI_Finalize();
      return -1;
    }
  }

  /* root node outputs parallelism information to screen */
  if (myid == 0)  printf(" starting MPI with %i processes\n",numprocs);

  /* root node sends n out to other processors */
  ierr = MPI_Bcast(&n, 1, MPI_INT, 0, MPI_COMM_WORLD);
  if (ierr != 0) {
    printf("Error in mpi_bcast = %i\n",ierr);
    ierr = MPI_Abort(MPI_COMM_WORLD, -1);
    return -1;
  }

  /* stop for illegal n */
  if (n < 1) {
    ierr = MPI_Finalize();
    return -1;
  }
  
  /* determine this processor's interval */
  int is = ((int) 1.0*n/numprocs)*myid;
  int ie = ((int) 1.0*n/numprocs)*(myid+1) - 1;
  if (myid == numprocs-1)  ie = n-1;

  /* root node allocates temperature field and solution arrays */
  double *T=NULL, *u=NULL, *v=NULL, *w=NULL;
  if (myid == 0) {
    T = malloc( n * sizeof(double) );
    u = malloc( n * sizeof(double) );
    v = malloc( n * sizeof(double) );
    w = malloc( n * sizeof(double) );

    /* set random temperature field, initial guesses at chemical densities */
    for (i=0; i<n; i++)  T[i] = random() / (pow(2.0,31.0) - 1.0);
    for (i=0; i<n; i++)  u[i] = 0.35;
    for (i=0; i<n; i++)  v[i] = 0.1;
    for (i=0; i<n; i++)  w[i] = 0.5;
  }
  /* other nodes allocate local temperature field and solution arrays */
  else {
    T = malloc( (ie-is+1) * sizeof(double) );
    u = malloc( (ie-is+1) * sizeof(double) );
    v = malloc( (ie-is+1) * sizeof(double) );
    w = malloc( (ie-is+1) * sizeof(double) );

    /* set initial guesses at chemical densities */
    for (i=0; i<(ie-is+1); i++)  u[i] = 0.35;
    for (i=0; i<(ie-is+1); i++)  v[i] = 0.1;
    for (i=0; i<(ie-is+1); i++)  w[i] = 0.5;
  }

  /* allocate gatherv/scatterv temporary arrays */
  int *counts = malloc( numprocs * sizeof(int) );
  int *displs = malloc( numprocs * sizeof(int) );

  /* start timer */
  double stime = MPI_Wtime();

  /* root node sends out portions of temperature field to all procs */
  int js, je;
  for (i=0; i<numprocs; i++) {
    js = floor(1.0*n/numprocs)*i;
    je = floor(1.0*n/numprocs)*(i+1) - 1;
    if (i == numprocs-1)  je = n-1;
    counts[i] = je-js+1;
    displs[i] = js;
  }
  ierr = MPI_Scatterv(T, counts, displs, MPI_DOUBLE, T, 
		      ie-is+1, MPI_DOUBLE, 0, MPI_COMM_WORLD);


  /* call solver over local intervals */
  int its, it;
  double res;
  for (i=0; i<(ie-is+1); i++) {
    it = is+i;
    chem_solver(T[i], &(u[i]), &(v[i]), &(w[i]), lam, eps, maxit, &its, &res);
    if (res < eps) {
/*       printf("    i = %i,  its = %i\n", i, its); */
    }
    else {
      printf("    error: i = %i,  its = %i,  res = %.2e,  u = %.2e,  v = %.2e,  w = %.2e\n", i, its, res, u[i], v[i], w[i]);
      ierr = MPI_Abort(MPI_COMM_WORLD, -1);
      return -1;
    }
  }

  /* root node collects results */
  ierr = MPI_Gatherv(u, ie-is+1, MPI_DOUBLE, u, counts, 
		     displs, MPI_DOUBLE, 0, MPI_COMM_WORLD);
  if (ierr != 0) {
    printf("Error in MPI_Gatherv = %i\n",ierr);
    ierr = MPI_Abort(MPI_COMM_WORLD, -1);
    return -1;
  }
  ierr = MPI_Gatherv(v, ie-is+1, MPI_DOUBLE, v, counts, 
		     displs, MPI_DOUBLE, 0, MPI_COMM_WORLD);
  if (ierr != 0) {
    printf("Error in MPI_Gatherv = %i\n",ierr);
    ierr = MPI_Abort(MPI_COMM_WORLD, -1);
    return -1;
  }
  ierr = MPI_Gatherv(w, ie-is+1, MPI_DOUBLE, w, counts, 
		     displs, MPI_DOUBLE, 0, MPI_COMM_WORLD);
  if (ierr != 0) {
    printf("Error in MPI_Gatherv = %i\n",ierr);
    ierr = MPI_Abort(MPI_COMM_WORLD, -1);
    return -1;
  }

  /* stop timer */
  double ftime = MPI_Wtime();
  double runtime = ftime - stime;

  /* output solution time */
  printf(" proc %i computed %i iterations\n", myid, ie-is+1);
  printf("     runtime = %.16e\n",runtime);

  /* free temperature and solution arrays */
  free(T);
  free(u);
  free(v);
  free(w);
  free(counts);
  free(displs);

  /* finalize MPI */
  ierr = MPI_Finalize();

} /* end main */

